File: LinkedList/SinglyLinkedList.py
# class of a node
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


# class of the signly linked list
class SinglyLinkedList:
    def __init__(self):
        self.head = None

    # function to calculate size of the list
    def size(self):
        # initialize temporary variable and size to zero
        current = self.head
        size = 0

        # count until current does not reach the end of the list i.e. NULL or None
        while current is not None:
            size += 1
            current = current.next
        return size

    # function to insert at the end of the list
    def insert_at_end(self, data):
        # Create the new node
        node = Node(data)

        # If the list is empty
        if self.head is None:
            self.head = node
        else:
            current = self.head

            # Otherwise move to the last node of the list
            while current.next is not None:
                current = current.next

            # Point the last node of the list to the new node
            # So that the new node gets added to the end of the list
            current.next = node

    # function to insert at the beginning of the list
    def insert_at_beg(self, data):
        node = Node(data)
        # Next pointer of the new node points to the head
        node.next = self.head

        # Updated the head as new node
        self.head = node

    # function to delete from the end of the list
    def delete_from_end(self):
        current = self.head
        previous = None

        if current is None:
            print("Linked List Underflow!!")
        else:
            while current.next is not None:
                previous = current
                current = current.next

            if previous is None:
                self.head = None
            else:
                previous.next = None

    # function to delete from the beginning of the list
    def delete_from_beg(self):
        current = self.head

        if current is None:
            print("Linked List Underflow!!")
        else:
            self.head = current.next

    # function to print the linked list data
    def print_data(self):
        current = self.head

        while current is not None:
            print(current.data, '->', end='')
            current = current.next
        print('End of list')


# Main program:
if __name__ == '__main__':
    # Create a singly linked list object
    linked_list = SinglyLinkedList()

    # Insert at the beginning 3, 2, 1
    linked_list.insert_at_beg(3)
    linked_list.insert_at_beg(2)
    linked_list.insert_at_beg(1)
    print('After insertion at the beginning:')
    linked_list.print_data()

    # Insert at the end of the list 4, 5
    linked_list.insert_at_end(4)
    linked_list.insert_at_end(5)
    print('After insertion at the end:')
    linked_list.print_data()

    # Delete 1 from the beginning
    print('After deletion at the beginning:')
    linked_list.delete_from_beg()
    linked_list.print_data()

    # Delete 5 from the end
    print('After deletion at the end:')
    linked_list.delete_from_end()
    linked_list.print_data()

    # print size of the list
    print("size: ", linked_list.size())

File: LinkedList/SinglyLinkedList.cpp
// Implementation of Singly linked list

#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* next;

        Node() {
            this->next = NULL;
        }

        Node(int data) {
            this->data = data;
            this->next = NULL;
        }
};

class LinkedList {
    public:
        Node* head = NULL;

        void insert_at_beg(int data){
            Node* temp = new Node(data);
            temp->next = this->head;
            this->head = temp;
        }

        void insert_at_end(int data){
            Node* itr = this->head;
            if (itr == NULL){
                this->insert_at_beg(data);
                return;
            }
            
            Node* temp = new Node(data);

            while(itr->next != NULL){
                itr = itr->next;
            }
            
            itr->next = temp;
        }
        

        void delete_from_beg(){
            if (this->head == NULL) return;

            Node* temp = this->head;
            this->head = this->head->next;
            free(temp);
        }

        void delete_from_end(){
            Node* itr = this->head;
            if (itr == NULL){
                return;
            }
            else if (itr->next == NULL){
                this->head = NULL;
                free(itr);
            }

            while(itr->next->next != NULL){
                itr = itr->next;
            }
            
            Node* temp = itr->next;
            itr->next = NULL;
            free(temp);
        }

        int size() {      
            Node* itr = this->head;
            int length = 0;

            while(itr != NULL){
                length += 1;
                itr = itr->next;
            }

            return length;
        }

        void print_data(){
            Node* itr = this->head;

            while(itr != NULL){
                cout << itr->data << " ";
                itr = itr->next;
            }
            cout << endl;
        }
};

int main() {
    LinkedList* ll = new LinkedList();
    
    ll->delete_from_beg();
    ll->delete_from_end();
    // Insert at the beginning 3, 2, 1
    ll->insert_at_beg(3);
    ll->insert_at_beg(2);
    ll->insert_at_beg(1);
    cout<<"After insertion at the beginning:\n";
    ll->print_data();

    // Insert at the end of the list 4, 5
    ll->insert_at_end(4);
    ll->insert_at_end(5);
    cout<<"After insertion at the end:\n";
    ll->print_data();

    // Delete 1 from the beginning
    cout<<"After deletion at the beginning:\n";
    ll->delete_from_beg();
    ll->print_data();

    // Delete 5 from the end
    cout<<"After deletion at the end:\n";
    ll->delete_from_end();
    ll->print_data();

    // print size of the list
    cout << "Size of the linked list: " << ll->size() << endl;
    return 0;
}



